File: akit/src/main/java/org/littletonrobotics/junction/wpilog/WPILOGReader.java
// Copyright (c) 2021-2025 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

package org.littletonrobotics.junction.wpilog;

import edu.wpi.first.util.datalog.DataLogIterator;
import edu.wpi.first.util.datalog.DataLogReader;
import edu.wpi.first.util.datalog.DataLogRecord;
import edu.wpi.first.wpilibj.DriverStation;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import org.littletonrobotics.junction.LogDataReceiver;
import org.littletonrobotics.junction.LogReplaySource;
import org.littletonrobotics.junction.LogTable;
import org.littletonrobotics.junction.LogTable.LogValue;
import org.littletonrobotics.junction.LogTable.LoggableType;

/** Replays log values from a WPILOG file. */
public class WPILOGReader implements LogReplaySource {
  private final String filename;
  private boolean isValid;

  private DataLogReader reader;
  private DataLogIterator iterator;

  private Long timestamp;
  private Map<Integer, String> entryIDs;
  private Map<Integer, LoggableType> entryTypes;
  private Map<Integer, String> entryCustomTypes;

  /**
   * Creates a new WPILOGReader.
   *
   * @param filename The log filename to read.
   */
  public WPILOGReader(String filename) {
    this.filename = filename;
  }

  public void start() {
    // Open log file
    try {
      reader = new DataLogReader(filename);
    } catch (IOException e) {
      DriverStation.reportError("[AdvantageKit] Failed to open replay log file.", true);
    }

    // Check validity
    if (!reader.isValid()) {
      DriverStation.reportError("[AdvantageKit] The replay log is not a valid WPILOG file.", false);
      isValid = false;
    } else if (!reader.getExtraHeader().equals(WPILOGConstants.extraHeader)) {
      DriverStation.reportError(
          "[AdvantageKit] The replay log was not produced by AdvantageKit.", true);
      isValid = false;
    } else {
      isValid = true;
    }

    // Create iterator and reset
    iterator = reader.iterator();
    timestamp = null;
    entryIDs = new HashMap<>();
    entryTypes = new HashMap<>();
    entryCustomTypes = new HashMap<>();
  }

  public boolean updateTable(LogTable table) {
    if (!isValid) {
      return false;
    }

    // Update timestamp (except the first cycle b/c it isn't known yet)
    if (timestamp != null) {
      table.setTimestamp(timestamp);
    }

    // Iterate over log
    boolean readError = false;
    while (iterator.hasNext()) {
      DataLogRecord record;
      try {
        record = iterator.next();
      } catch (Exception e) {
        readError = true;
        break;
      }

      if (record.isControl()) {
        if (record.isStart()) { // Ignore other control records
          entryIDs.put(record.getStartData().entry, record.getStartData().name);
          String typeStr = record.getStartData().type;
          entryTypes.put(record.getStartData().entry, LoggableType.fromWPILOGType(typeStr));
          if (typeStr.startsWith("proto:")
              || typeStr.startsWith("struct:")
              || typeStr.equals("structschema")) {
            entryCustomTypes.put(record.getStartData().entry, typeStr);
          }
        }

      } else {
        String entry = entryIDs.get(record.getEntry());
        if (entry != null) {
          if (entry.equals(LogDataReceiver.timestampKey)) {
            boolean firstTimestamp = timestamp == null;
            timestamp = record.getInteger();
            if (firstTimestamp) {
              table.setTimestamp(timestamp);
            } else {
              break; // End of cycle
            }

          } else if (timestamp != null && record.getTimestamp() == timestamp) {
            entry = entry.substring(1); // Remove leading slash
            if (entry.startsWith("ReplayOutputs")) {
              // Don't retrieve old replay outputs
              continue;
            }
            String customType = entryCustomTypes.get(record.getEntry());
            switch (entryTypes.get(record.getEntry())) {
              case Raw:
                table.put(entry, new LogValue(record.getRaw(), customType));
                break;
              case Boolean:
                table.put(entry, new LogValue(record.getBoolean(), customType));
                break;
              case Integer:
                table.put(entry, new LogValue(record.getInteger(), customType));
                break;
              case Float:
                table.put(entry, new LogValue(record.getFloat(), customType));
                break;
              case Double:
                table.put(entry, new LogValue(record.getDouble(), customType));
                break;
              case String:
                table.put(entry, new LogValue(record.getString(), customType));
                break;
              case BooleanArray:
                table.put(entry, new LogValue(record.getBooleanArray(), customType));
                break;
              case IntegerArray:
                table.put(entry, new LogValue(record.getIntegerArray(), customType));
                break;
              case FloatArray:
                table.put(entry, new LogValue(record.getFloatArray(), customType));
                break;
              case DoubleArray:
                table.put(entry, new LogValue(record.getDoubleArray(), customType));
                break;
              case StringArray:
                table.put(entry, new LogValue(record.getStringArray(), customType));
                break;
            }
          }
        }
      }
    }

    // Continue if there is more data
    return iterator.hasNext() && !readError;
  }
}

File: akit/src/main/native/cpp/wpilog/WPILOGReader.cpp
#include "akit/wpilog/WPILOGReader.h"
#include "akit/wpilog/WPILOGConstants.h"
#include "akit/LogDataReceiver.h"

using namespace akit::wpilog;

void WPILOGReader::start() {
    reader = wpi::log::DataLogReader{wpi::MemoryBuffer::GetFile(filename).value()};
    if (!reader->IsValid()) {
        FRC_ReportError(frc::err::Error, "[AdvantageKit] The replay log is not a valid WPILOG file.");
        isValid = false;
    } else if (reader->GetExtraHeader() == WPILOGConstants::EXTRA_HEADER) {
        FRC_ReportError(frc::err::Error, "[AdvantageKit] The replay log was not produced by AdvantageKit.");
        isValid = false;
    } else isValid = true;

    iterator = reader->begin();
}

bool WPILOGReader::updateTable(LogTable&& table) {
    if (!isValid) return false;

    if (timestamp)
        table.setTimestamp(*timestamp);
    
    bool readError = false;
    for (; *iterator < reader->end(); (*iterator)++) {
        auto& record = **iterator;
        if (record.IsControl()) {
            if (record.IsStart()) {
                wpi::log::StartRecordData startRecord;
                record.GetStartData(&startRecord);
                entryIDs[startRecord.entry] = startRecord.name;
                LogTable::LoggableType loggableType{std::distance(LogTable::WPILOG_TYPES.begin(), std::ranges::find(LogTable::WPILOG_TYPES, startRecord.type))};
                entryTypes[startRecord.entry] = loggableType;
                if ((loggableType == LogTable::LoggableType::Raw && startRecord.type != "raw") || startRecord.type == "json")
                    entryCustomTypes[startRecord.entry] = startRecord.type;
            }
        } else {
            auto entry = entryIDs.find(record.GetEntry());
            if (entry != entryIDs.end()) {
                if (entry->second == LogDataReceiver::TIMESTAMP_KEY) {
                    bool firstTimestamp = timestamp.has_value();
                    int64_t time;
                    record.GetInteger(&time);
                    timestamp = time;
                    if (firstTimestamp)
                        table.setTimestamp(*timestamp);
                    else break;
                } else if (timestamp && record.GetTimestamp() == timestamp) {
                    entry->second = entry->second.substr(1);
                    if (entry->second.starts_with("ReplayOutputs"))
                        continue;
                    std::string customType = entryCustomTypes[record.GetEntry()];
                    switch (entryTypes[record.GetEntry()]) {
                        case LogTable::LoggableType::Raw: {
                            auto value = record.GetRaw();
                            std::byte* begin = static_cast<std::byte*>(value.data());
                            table.put(entry->second, LogTable::LogValue{std::vector<std::byte>{value.data(), value.end()}, customType});
                            break;
                        } case LogTable::LoggableType::Boolean: {
                            bool value;
                            record.GetBoolean(&value);
                            table.put(entry->second, LogTable::LogValue{value, customType});
                            break;
                        } case LogTable::LoggableType::Integer: {
                            int64_t value;
                            record.GetInteger(&value);
                            table.put(entry->second, LogTable::LogValue{static_cast<long>(value), customType});
                            break;
                        } case LogTable::LoggableType::Float: {
                            float value;
                            record.GetFloat(&value);
                            table.put(entry->second, LogTable::LogValue{value, customType});
                            break;
                        } case LogTable::LoggableType::Double: {
                            double value;
                            record.GetDouble(&value);
                            table.put(entry->second, LogTable::LogValue{value, customType});
                            break;
                        } case LogTable::LoggableType::String: {
                            std::string_view value;
                            record.GetString(&value);
                            table.put(entry->second, LogTable::LogValue{std::string{value}, customType});
                            break;
                        } case LogTable::LoggableType::BooleanArray: {
                            std::vector<int> value;
                            record.GetBooleanArray(&value);
                            table.put(entry->second, LogTable::LogValue{std::vector<bool>{value.begin(), value.end()}, customType});
                            break;
                        } case LogTable::LoggableType::IntegerArray: {
                            std::vector<int64_t> value;
                            record.GetIntegerArray(&value);
                            table.put(entry->second, LogTable::LogValue{std::vector<long>{value.begin(), value.end()}, customType});
                            break;
                        } case LogTable::LoggableType::FloatArray: {
                            std::vector<float> value;
                            record.GetFloatArray(&value);
                            table.put(entry->second, LogTable::LogValue{value, customType});
                            break;
                        } case LogTable::LoggableType::DoubleArray: {
                            std::vector<double> value;
                            record.GetDoubleArray(&value);
                            table.put(entry->second, LogTable::LogValue{value, customType});
                            break;
                        } case LogTable::LoggableType::StringArray: {
                            std::vector<std::string_view> value;
                            record.GetStringArray(&value);
                            table.put(entry->second, LogTable::LogValue{std::vector<std::string>{value.begin(), value.end()}, customType});
                            break;
                        }
                    }
                }
            }
        }

        return *iterator < reader->end() && !readError;
    }
}
